runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -973,65 +1302,312 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
 	Assert_MM_true(edenMaximumCount >= 1);
 	Assert_MM_true(edenMaximumCount >= edenMinimumCount);
-	
-	UDATA desiredEdenCount = freeRegions;
-	if (desiredEdenCount > edenMaximumCount) {
-		desiredEdenCount = edenMaximumCount;
-	} else if (desiredEdenCount < edenMinimumCount) {
-		desiredEdenCount = edenMinimumCount;
-	}
+
+	/* Allow eden to expand as much as it wants, as long as the total heap can expand to accomodate it */
+	uintptr_t desiredEdenCount = OMR_MAX(edenMaximumCount, edenMinimumCount);
+	intptr_t desiredEdenChangeSize = (intptr_t)desiredEdenCount - (intptr_t)_edenRegionCount;
+	/* Determine if eden should try to grow anyways, or if the heap is tight on memory */
+	uintptr_t maximumHeap = _extensions->softMx == 0 ? _extensions->memoryMax : _extensions->softMx;
+	uintptr_t maximumHeapRegions = maximumHeap / _regionManager->getRegionSize();
+	intptr_t maxHeapExpansionRegions = OMR_MAX(0, ((intptr_t)maximumHeapRegions - (intptr_t)_numberOfHeapRegions - 1));
+
 	Trc_MM_SchedulingDelegate_calculateEdenSize_dynamic(env->getLanguageVMThread(), desiredEdenCount, _edenSurvivalRateCopyForward, _nonEdenSurvivalCountCopyForward, freeRegions, edenMinimumCount, edenMaximumCount);
-	if (desiredEdenCount <= freeRegions) {
-		_edenRegionCount = desiredEdenCount;
+
+	/* Make sure that the total heap can expand enough to satisfy the desired change in eden size
+	 * If heap is fully expanded (or close to) make sure that there are enough free regions to satisfy given eden size change 
+	 */
+	intptr_t maxEdenChange = 0;
+
+	if (0 == maxHeapExpansionRegions) {
+		/* 
+		 * The heap is fully expanded. Eden will be stealing free regions from the entire heap, without telling the heap to grow. 
+		 * Note: When heap is fully expanded, the eden sizing logic knows how much free memory is available in the heap, and knows to not grow too much 
+		 */
+		maxEdenChange = freeRegions;
+		_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = 0;
 	} else {
-		/* there isn't enough memory left for a desired Eden. Allow Eden to shrink to free size(could be less than minimum size or 0) before
-		 * triggering an allocation failure collection (i.e. a global STW collect)
-		 */ 
-		_edenRegionCount = freeRegions;
-		Trc_MM_SchedulingDelegate_calculateEdenSize_reduceToFreeBytes(env->getLanguageVMThread(), desiredEdenCount, _edenRegionCount);
+		/* Eden will inform the total heap resizing logic, that it needs to change total heap size in order to maintain same "tenure" size */
+		maxEdenChange = maxHeapExpansionRegions;
+		intptr_t edenChangeWithSurvivorHeadroom = desiredEdenChangeSize;
+		/* Total heap needs to be aware that by changing eden size, the amount of survivor space might also need to change */
+		if (0 < desiredEdenChangeSize) {
+			edenChangeWithSurvivorHeadroom = desiredEdenChangeSize + (intptr_t)ceil(((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward));
+		} else if ((0 > desiredEdenChangeSize) && (16 < _edenRegionCount)) {
```

Comment:
There might be  (although rare) configurations where max regions counts is much smaller or bigger then default 1000-2000. Perhaps we can express this comparison with 16 in relative terms like:

_edenRegionCount * 64 > maxRegionCount 

Although the whole idea of special casing for startup with some hard coded ratio  representing 'small region' isn't ideal...

Reply:
`_edenRegionCount * 64 > maxRegionCount` seems reasonable!

The only thing that this threshold is really doing, is including/excluding survivor regions when informing the total heap by how many regions it needs to change. Through experimentation, 16 was reasonable number of eden regions to start to include for the reduction of survivor regions needed, but the alternative you mentioned would also work just fine. Having some sort of case for when eden is very small (whether relative to maxRegions or some absolute value) just prevents the heap from shrinking more than it needs to

Reply:
This has been fixed as of latest commit, and it behaves as expected. 
With very small heaps (tested with -Xmx32M) the condition allows eden to include survivor headroom essentially right away. (threshold basically becomes true when eden count is 1), but when 1000-2000 regions are used (the common case), this condition behaves the same was as before (anywhere from ~ 16-32 eden regions is now the threshold)